ext-dynamic-commitments.md

```diff
@@ -0,0 +1,1087 @@
+current channel state machine. Assuming an agreement can be reached, we will
+proceed to the execution phase. During the execution phase, we apply the updates
+to the channel state machine, exchanging the necessary information to be able to
+apply those updates.
+
+## Proposal Phase
+
+As a prerequisite to the proposal phase of a Dynamic Commitment negotiation, the
+channel must be in a [quiesced](https://github.com/lightning/bolts/pull/869)
+state.
+
+### Node Roles
+
+In every dynamic commitment negotiation, there are two roles: the `initiator`
+and the `responder`. It is necessary for both nodes to agree on which node is
+the `initiator` and which node is the `responder`. This is important because if
+the dynamic commitment negotiation results in a re-anchoring step (described
+later), it is the `initiator` that is responsible for paying the fees for the
+kickoff transaction. The `initiator` is determined by who has the `initiator`
+role established by the quiescence process.
+
+### Negotiation TLVs
+
+The following TLVs are used throughout the negotiation phase of the protocol
+and are common to all messages in the negotiation phase.
+
+#### dust_limit_satoshis
+
+- type: 0
+  data:
+    * [`u64`:`dust_limit_satoshis`]
+
+#### max_htlc_value_in_flight_msat
+
+- type: 1
+  data:
+    * [`u64`:`senders_max_htlc_value_in_flight_msat`]
+
+#### channel_reserve_satoshis
+
+- type: 2
+  data:
+    * [`u64`:`recipients_channel_reserve_satoshis`]
+
+#### to_self_delay
+
+- type: 3
+  data:
+    * [`u16`:`recipients_to_self_delay`]
+
+#### max_accepted_htlcs
+
+- type: 4
+  data:
+    * [`u16`:`senders_max_accepted_htlcs`]
+
+#### funding_pubkey
+
+- type: 5
+  data:
+    * [`point`:`senders_funding_pubkey`]
+
+#### channel_type
+
+- type: 6
+  data:
+    * [`...*byte`:`channel_type`]
+
+#### kickoff_feerate_per_kw
+
+- type: 7
+  data:
+    * [`u32`:`kickoff_feerate_per_kw`]
+
+
+### Proposal Messages
+
+Three new messages are introduced that are common to all dynamic commitment
+flows. They let each channel party propose which channel parameters they wish to
+change as well as accept or reject the proposal made by their counterparty.
+
+#### `dyn_propose`
+
+This message is sent to negotiate the parameters of a dynamic commitment
+upgrade. The overall protocol flow is depicted below. This message is always
+sent by the `initiator`.
+
+        +-------+                               +-------+
+        |       |--(1)------ dyn_propose ------>|       |
+        |   A   |                               |   B   |
+        |       |<-(4)---{dyn_ack|dyn_reject}---|       |
+        +-------+                               +-------+
+
+1. type: 111 (`dyn_propose`)
+2. data:
+   * [`32*byte`:`channel_id`]
+   * [`dyn_propose_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `dyn_propose_tlvs`
+2. types:
+    1. type: 0 (`dust_limit_satoshis`)
+    2. data:
+        * [`u64`:`dust_limit_satoshis`]
+    1. type: 1 (`max_htlc_value_in_flight_msat`)
```

Comment:
Shouldn't all of these TLVs be even? Under okay to be odd reasoning, the receiving node can receive and ignore all of the odd fields here even if they don't understand what they are which could lead to miscommunication between nodes.

Reply:
Yeah. This is a good point.

I have been trying to figure out exactly when IOKTBO applies as there are a number of "core protocol messages" with odd message numbers, but yes, iiuc this is never the case with TLVs and this is a mistake. The choice to do it was naively more to keep the rejection vector compact since limiting to even numbers only makes it sparse, but this is probably a poor tradeoff since even though it's a 2x increase we're literally talking about a *single byte* 😬.

I will fix this in the next revision. Thank you for pointing it out!